src/Surfnet/StepupSelfService/SelfServiceBundle/Controller/SamlController.php

```diff
@@ -18,29 +18,110 @@
 
 namespace Surfnet\StepupSelfService\SelfServiceBundle\Controller;
 
-use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;
+use Exception;
 use Surfnet\SamlBundle\Http\XMLResponse;
-use Symfony\Bundle\FrameworkBundle\Controller\Controller as FrameworkController;
+use Surfnet\SamlBundle\SAML2\Response\Assertion\InResponseTo;
+use Surfnet\StepupBundle\Value\SecondFactorType;
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
 
-class SamlController extends FrameworkController
+class SamlController extends Controller
 {
     /**
-     * @Template
+     * @param string $secondFactorId
+     *
+     * @return \Symfony\Component\HttpFoundation\RedirectResponse
+     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
+     * @throws \Symfony\Component\Security\Core\Exception\AccessDeniedException
      */
+    public function testSecondFactorAction($secondFactorId)
+    {
+        $this->get('logger')->notice(
+            'Starting second factor test'
+        );
+
+        $secondFactorService = $this->get('surfnet_stepup_self_service_self_service.service.second_factor');
+        $identity            = $this->getIdentity();
+
+        if (!$secondFactorService->identityHasSecondFactorOfStateWithId($identity->id, 'vetted', $secondFactorId)) {
+            $this->get('logger')->error(
+                sprintf(
+                    'Identity "%s" tried to test second factor "%s", but does not own that second factor or it is not vetted',
+                    $identity->id,
+                    $secondFactorId
+                )
+            );
+
+            throw $this->createNotFoundException();
```

Comment:
This is just a Symfony convenience method that creates a NotFoundHttpException with a default message, I guess it is a matter of personal preference.